internal/controller/observedobjectcollection/reconciler.go

```diff
@@ -0,0 +1,318 @@
+			po.SetLabels(l)
+		} else {
+			po.SetLabels(map[string]string{membershipLabelKey: mlv})
+		}
+		if err := r.client.Patch(ctx, po, client.Apply, fieldOwner, client.ForceOwnership); err != nil {
+			werr := errors.Wrap(err, "cannot create observed object")
+			c.Status.SetConditions(xpv1.ReconcileError(werr))
+			_ = r.client.Status().Update(ctx, c)
+			return ctrl.Result{}, werr
+		}
+
+		log.Debug("created observed object", "name", po.GetName())
+		refs.Insert(v1alpha1.ObservedObjectReference{Name: name})
+	}
+
+	// Remove collection members that either do not exist anymore or are no match.
+	ol := &v1alpha2.ObjectList{}
+	if err := r.client.List(ctx, ol, client.MatchingLabels(map[string]string{membershipLabelKey: mlv})); err != nil {
+		werr := errors.Wrapf(err, "cannot list members matching label %v=%v", membershipLabelKey, mlv)
+		c.Status.SetConditions(xpv1.ReconcileError(werr))
+		_ = r.client.Status().Update(ctx, c)
+		return ctrl.Result{}, werr
+	}
+
+	oldRefs := sets.New[v1alpha1.ObservedObjectReference](c.Status.ObjectRefs...)
+	for i := range ol.Items {
+		name := ol.Items[i].Name
+		if refs.Has(v1alpha1.ObservedObjectReference{Name: name}) {
+			continue
+		}
+		log.Debug("Removing", "name", name)
+		if err := r.client.Delete(ctx, &ol.Items[i]); err != nil {
+			werr := errors.Wrapf(err, "cannot delete observed object %s", name)
+			c.Status.SetConditions(xpv1.ReconcileError(werr))
+			_ = r.client.Status().Update(ctx, c)
+			return ctrl.Result{}, werr
+		}
+	}
+	c.Status.SetConditions(xpv1.ReconcileSuccess(), xpv1.Available())
+
+	// Update references only if more or less objects got matched since the last reconciliation.
+	if !refs.Equal(oldRefs) {
+		c.Status.ObjectRefs = refs.UnsortedList()
+	}
+	c.Status.MembershipLabel = map[string]string{membershipLabelKey: mlv}
+
+	return ctrl.Result{RequeueAfter: r.pollInterval()}, r.client.Status().Update(ctx, c)
+}
+
+func observedObjectName(collection client.Object, matchedObject client.Object) (string, error) {
+	name := fmt.Sprintf("%s-%s-%s-%s", collection.GetName(), strings.ToLower(matchedObject.GetObjectKind().GroupVersionKind().Kind), matchedObject.GetNamespace(), matchedObject.GetName())
+	// If the name length is less than 253 chars,
+	// we can use it as-is.
+	// https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#dns-subdomain-names
+	if len(name) <= 253 {
+		return name, nil
+	}
+
+	// Otherwise, compute md5 hash of it, to reduce it length.
+	h := md5.New() //#nosec G401 -- used only for unique name generation
+	if _, err := h.Write([]byte(name)); err != nil {
+		return "", err
+	}
+	id, err := uuid.FromBytes(h.Sum(nil))
+	return id.String(), err
+}
+
+func membershipLabelValue(collection *v1alpha1.ObservedObjectCollection) (string, error) {
+	if len(collection.GetName()) <= 63 {
+		return collection.GetName(), nil
+	}
+	// Otherwise, compute md5 hash of it, to reduce it length.
+	h := md5.New() //#nosec G401 -- used only for unique name generation
```

Comment:
Just checked how Crossplane doing it with `crossplane.io/composite` label: https://github.com/crossplane/crossplane/blob/d35ecef8b3d9094261dbda64e72498cda1520797/internal/controller/apiextensions/composite/api.go#L401

Looks like we don't have special handling for long names rather try to add it as label, it fails and we return the error to the user so that they can adjust their naming. I remember I observed this while applying some composites (of platform-refs) and used shorter names with my compositions. 

I would agree that this is not the best experience, but still would follow upstream implementation here as well. To keep the logic here simple and not have cryptic owner labels when they are long.